File: aoc2022/kotlin/src/main/kotlin/Day05.kt
import util.AdventOfCode
import java.util.*

object Day05 : AdventOfCode() {
    data class CargoArea(
        val stacks: Map<Int, LinkedList<Char>>,
        val instructions: Sequence<Instruction>,
        val crane: Crane
    ) {
        fun processInstructions() = instructions.forEach { crane.work(stacks, it) }
        fun topBoxLabels() = stacks.toSortedMap().values.map { it.first() }.joinToString("")
    }

    interface Crane {
        fun work(stacks: Map<Int, LinkedList<Char>>, instruction: Instruction)
    }

    object CrateMover9000 : Crane {
        override fun work(stacks: Map<Int, LinkedList<Char>>, instruction: Instruction) {
            transfer(instruction.numToMove, stacks[instruction.toCol]!!, stacks[instruction.fromCol]!!)
        }
    }

    object CrateMover9001 : Crane {
        private val tempStack = LinkedList<Char>()
        override fun work(stacks: Map<Int, LinkedList<Char>>, instruction: Instruction) {
            transfer(instruction.numToMove, tempStack, stacks[instruction.fromCol]!!)
            transfer(instruction.numToMove, stacks[instruction.toCol]!!, tempStack)
        }
    }

    data class Instruction(val numToMove: Int, val fromCol: Int, val toCol: Int) {
        constructor(n: String, f: String, t: String) : this(n.toInt(), f.toInt(), t.toInt())
    }

    private fun String.toInstruction() =
        """move (\d+) from (\d+) to (\d+)""".toRegex()
            .matchEntire(this)!!.groupValues
            .let { Instruction(it[1], it[2], it[3]) }

    private fun parse(input: String, crane: Crane): CargoArea {
        val (diagram, instrInput) = input.split("\n\n").map { it.lines() }

        val cols = diagram.last().trim().split("   ").map {
            it.toInt() to (it.toInt() - 1) * 4 + 1
        }

        val stacks = cols.associate { it.first to LinkedList<Char>() }
        diagram.dropLast(1).reversed().forEach { line ->
            cols.filter { (_, col) -> line[col] != ' ' }.forEach { (id, col) -> stacks[id]?.push(line[col]) }
        }

        return CargoArea(stacks, instrInput.asSequence().map { it.toInstruction() }, crane)
    }

    private fun transfer(n: Int, a: LinkedList<Char>, b: LinkedList<Char>) = repeat(n) { a.push(b.pop()) }

    fun part1(input: String): String {
        val cargoArea = parse(input, CrateMover9000)
        cargoArea.processInstructions()
        return cargoArea.topBoxLabels()
    }

    fun part2(input: String): String {
        val cargoArea = parse(input, CrateMover9001)
        cargoArea.processInstructions()
        return cargoArea.topBoxLabels()
    }

    @JvmStatic
    fun main(args: Array<String>) {
        println("Day 5")
        println("\tPart 1: ${part1(inputFileString)}")
        println("\tPart 2: ${part2(inputFileString)}")
    }
}

File: aoc2022/fsharp/Days/Day05.fs
module Day05

open System
open System.Text.RegularExpressions
open FSharpx.Text

type Instruction = { N: int; From: int; To: int }

let replace (m: Map<int, char list>) (fromCol: int) (toCol: int) ((fromList, toList): char list * char list) =
    m |> Map.add fromCol fromList |> Map.add toCol toList

let transfer' (num: int) (fromCol: char list) (toCol: char list) =
    let rec loop n f t =
        if (n > 0) then
            let newFrom = (f |> List.skip 1)
            let newTo = List.Cons(f.Head, t)
            loop (n - 1) newFrom newTo
        else
            (f, t)

    loop num fromCol toCol

let transfer (n: int) (fromCol: char list option) (toCol: char list option) =
    match fromCol, toCol with
    | Some f, Some t -> transfer' n f t
    | _, _ -> failwith "Unknown error invoking transfer"


type Crane =
    | CrateMover9000
    | CrateMover9001

    member this.DoWork (stacks: Map<int, char list>) (instruction: Instruction) =
        match this with
        | CrateMover9000 ->
            transfer instruction.N (stacks.TryFind instruction.From) (stacks.TryFind instruction.To)
            |> replace stacks instruction.From instruction.To
        | CrateMover9001 ->
            let fromCol, tempStack =
                transfer instruction.N (stacks.TryFind instruction.From) (Some <| List.empty)

            let _, toCol =
                transfer instruction.N (Some tempStack) (stacks.TryFind instruction.To)

            (fromCol, toCol) |> replace stacks instruction.From instruction.To


type CargoArea =
    { Stacks: Map<int, char list>
      Instructions: Instruction seq
      Crane: Crane }

let instrRegex = Regex(@"move (\d+) from (\d+) to (\d+)", RegexOptions.Compiled)
let colRegex = Regex(@"\d ", RegexOptions.Compiled)

let toInstruction (input: string) =
    let m = (input |> instrRegex.Match)

    { N = (int m.Groups[1].Value)
      From = (int m.Groups[2].Value)
      To = (int m.Groups[3].Value) }

let stackExtractor (col: int) (lines: string seq) =
    lines
    |> Seq.rev
    |> Seq.skip 1
    |> Seq.map (fun line -> line[col])
    |> Seq.filter (fun c -> c <> ' ')
    |> Seq.rev
    |> Seq.toList

let toKVTuples (lines: string seq) (label: int, col: int) : int * char list = (label, stackExtractor col lines)

let toStacks (diagram: string seq) =
    seq {
        for m in colRegex.Matches(diagram |> Seq.last) do
            yield m.Value.Trim() |> int, m.Index
    }
    |> Seq.map (toKVTuples diagram)
    |> Map.ofSeq

let partitionInput (input: string) =
    let inputParts =
        input.Split([| "\n\n" |], StringSplitOptions.RemoveEmptyEntries)
        |> Array.map Strings.toLines
        |> Array.take 2

    if inputParts |> Array.length <> 2 then
        failwith "inputParts has less than 2 elements"
    else
        inputParts[0], inputParts[1]

let parse (input: string) (crane: Crane) : CargoArea =
    let diagram, instrInput = partitionInput input

    { Stacks = toStacks diagram
      Instructions = (instrInput |> Seq.map toInstruction)
      Crane = crane }

let processInstructions (cargoArea: CargoArea) =
    cargoArea.Instructions |> Seq.fold cargoArea.Crane.DoWork cargoArea.Stacks

let part1 (input: string) =
    (parse input CrateMover9000)
    |> processInstructions
    |> Map.values
    |> Seq.map (fun cs -> (cs |> Seq.head).ToString())
    |> String.concat ""

let part2 (input: string) =
    (parse input CrateMover9001)
    |> processInstructions
    |> Map.values
    |> Seq.map (fun cs -> (cs |> Seq.head).ToString())
    |> String.concat ""
